src/Nest/Document/Multiple/BulkAll/BulkAllObservable.cs

```diff
@@ -0,0 +1,219 @@
+					}
+					observer.OnCompleted();
+					break;
+				case TaskStatus.Faulted:
+					observer.OnError(task.Exception.InnerException);
+					break;
+				case TaskStatus.Canceled:
+					observer.OnError(new TaskCanceledException(task));
+					break;
+			}
+		}
+
+		private async Task<IBulkAllResponse> BulkAsync(IList<T> buffer, long page, int backOffRetries)
+		{
+			this._compositeCancelToken.ThrowIfCancellationRequested();
+
+			var r = this._partionedBulkRequest;
+			var response = await this._client.BulkAsync(s =>
+			{
+				s.IndexMany(buffer).Index(r.Index).Type(r.Type);
+				if (!string.IsNullOrEmpty(r.Pipeline)) s.Pipeline(r.Pipeline);
+				if (r.Refresh) s.Refresh(r.Refresh);
+				if (!string.IsNullOrEmpty(r.Routing)) s.Pipeline(r.Routing);
+				if (r.Consistency.HasValue) s.Consistency(r.Consistency.Value);
+				return s;
+			}, this._compositeCancelToken);
+
+			this._compositeCancelToken.ThrowIfCancellationRequested();
+			if (!response.IsValid && backOffRetries < this._backOffRetries)
+			{
+				this._incrementRetries();
+				//wait before or after fishing out retriable docs?
+				await Task.Delay(this._backOffTime);
+				var retryDocuments = response.Items.Zip(buffer, (i, d) => new { i, d })
+					.Where(x => x.i.Status == 429)
+					.Select(x => x.d)
+					.ToList();
+
+				return await this.BulkAsync(retryDocuments, page, ++backOffRetries);
+			}
+			else if (!response.IsValid)
+			{
+				this._incrementFailed();
+				throw Throw($"Bulk indexing failed and after retrying {backOffRetries} times", response.ApiCall);
+			}
+			return new BulkAllResponse { Retries = backOffRetries, Page = page };
+		}
+
+		private sealed class PartitionHelper<TDocument> : IEnumerable<IList<TDocument>>
+		{
+			readonly IEnumerable<TDocument> _items;
+			readonly int _partitionSize;
+			readonly int _semaphoreSize;
+			bool _hasMoreItems;
+
+			internal PartitionHelper(IEnumerable<TDocument> i, int ps, int semaphoreSize)
+			{
+				_items = i;
+				_semaphoreSize = semaphoreSize;
+				_partitionSize = ps;
+			}
+
+			IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
+			public IEnumerator<IList<TDocument>> GetEnumerator()
+			{
+				using (var enumerator = _items.GetEnumerator())
+				{
+					_hasMoreItems = enumerator.MoveNext();
+					while (_hasMoreItems)
+						yield return GetNextBatch(enumerator).ToList();
+				}
+			}
+
+			IEnumerable<TDocument> GetNextBatch(IEnumerator<TDocument> enumerator)
+			{
+				for (int i = 0; i < _partitionSize; ++i)
+				{
+					yield return enumerator.Current;
+					_hasMoreItems = enumerator.MoveNext();
+					if (!_hasMoreItems) yield break;
+				}
+			}
+
+			public Task ForEachAsync<TResult>(
+				Func<IList<TDocument>, long, Task<TResult>> taskSelector,
+				Action<IList<TDocument>, TResult> resultProcessor,
+				Action<Task> done
+			)
+			{
+				var semaphore = new SemaphoreSlim(initialCount: _semaphoreSize, maxCount: _semaphoreSize);
```

Comment:
I'm already using a stephen toub variant :+1:  The method you link relies on the reader to be choped up into `dop` (degree of parallelism) and then itterate over these chunks and do the work on that. 

Won't work with 10.000.000 record dataset chopped into 4 concurrent unit of works. 

In our case the writer (to elasticsearch) is concurrent and whoever feeds us the `IEnumerable` does not have to make any guarantees on the thread safety of the collection or make sure it can witstand calling MoveNext too many times (as would be the case by using a double generator completely lazy approach). 

In *most* cases i reckon the real concurrency benefit is not reading from source (xml, db record set, files) in parallel but parallize the bulk indexing to elasticsearch. 

If it is, folks can still create their own more explicit concurrent read implementation behind the `IEnumerable<T>`